### ISSUE TITLE
Keyring file does not exist. Please try again.

### ISSUE THREAD

**Issue by ilium007** [boredazfcuk/docker-icloudpd#395]
Not sure what's happening here. 

Created with:

`docker create --name icloudpd --hostname icloudpd --network icloudpd_bridge --restart=always --env TZ=Australia/Sydney --volume /volume1/BACKUP01/icloudpd/config:/config --volume /volume1/BACKUP01/icloudpd/images:/images   boredazfcuk/icloudpd`

In the logs I see:

```
2023-09-27 10:00:34 INFO     Directory is writable: /config/python_keyring/
2023-09-27 10:00:34 INFO     Directory is writable: /home/user/.local/share/
```

but the container is looking for python_keyring in `/home/user/.local/share/python_keyring`

```
FileExistsError: [Errno 17] File exists: '/home/user/.local/share/python_keyring'
2023-09-27 10:00:51 ERROR    Keyring file does not exist. Please try again.
```

```
xxxx@synology01:~$ docker exec -it icloudpd sync-icloud.sh --Initialise
Password:

2023-09-27 10:00:33 INFO     ***** boredazfcuk/icloudpd container for icloud_photo_downloader v1.0.635 started *****
2023-09-27 10:00:33 INFO     ***** For support, please go here: https://github.com/boredazfcuk/docker-icloudpd *****
2023-09-27 10:00:33 INFO     Alpine Linux 3.18.3
2023-09-27 10:00:33 INFO     Python version: 3.11.5
2023-09-27 10:00:34 INFO     Loading configuration from: /config/icloudpd.conf
2023-09-27 10:00:34 INFO     Apple ID: xxxx@mac.com
2023-09-27 10:00:34 INFO     Authentication Type: MFA
2023-09-27 10:00:34 INFO     Cookie path: /config/xxxx
2023-09-27 10:00:34 INFO     Cookie expiry notification period: 7
2023-09-27 10:00:34 INFO     Download destination directory: /images
2023-09-27 10:00:34 INFO     Folder structure: {:%Y/%m/%d}
2023-09-27 10:00:34 INFO     Synchronisation interval: 86400
2023-09-27 10:00:34 INFO     Synchronisation delay (minutes): 0
2023-09-27 10:00:34 INFO     Set EXIF date/time: false
2023-09-27 10:00:34 INFO     Auto delete: false
2023-09-27 10:00:34 INFO     Delete after download: false
2023-09-27 10:00:34 INFO     Photo size: original
2023-09-27 10:00:34 INFO     Single pass mode: false
2023-09-27 10:00:34 INFO     Skip download check: false
2023-09-27 10:00:34 INFO     Skip live photos: false
2023-09-27 10:00:34 INFO     Number of most recently added photos to download: Download All Photos
2023-09-27 10:00:34 INFO     Downloading photos from album: Download All Photos
2023-09-27 10:00:34 INFO     Stop downloading when prexisiting files count is: Download All Photos
2023-09-27 10:00:34 INFO     Live photo size: original
2023-09-27 10:00:34 INFO     Skip videos: false
2023-09-27 10:00:34 INFO     Convert HEIC to JPEG: false
2023-09-27 10:00:34 INFO     Downloading from: icloud.com
2023-09-27 10:00:34 INFO     Authentication domain: com
2023-09-27 10:00:34 INFO     Ignore Synology extended attribute directories: Disabled
2023-09-27 10:00:34 INFO     Script launch parameters: --Initialise
2023-09-27 10:00:34 INFO     Directory is writable: /config/python_keyring/
2023-09-27 10:00:34 INFO     Directory is writable: /home/user/.local/share/
Enter iCloud password for xxxx@mac.com:
Save password in keyring?  [y/N]: y
Traceback (most recent call last):
  File "/opt/icloudpd_latest/bin/icloud", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/cmdline.py", line 219, in main
    utils.store_password_in_keyring(username, password)
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/utils.py", line 53, in store_password_in_keyring
    return keyring.set_password(
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/keyring/core.py", line 60, in set_password
    get_keyring().set_password(service_name, username, password)
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/keyrings/alt/file_base.py", line 129, in set_password
    self._write_config_value(service, username, password_base64)
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/keyrings/alt/file_base.py", line 137, in _write_config_value
    self._ensure_file_path()
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/keyrings/alt/file_base.py", line 163, in _ensure_file_path
    os.makedirs(storage_root)
  File "<frozen os>", line 225, in makedirs
FileExistsError: [Errno 17] File exists: '/home/user/.local/share/python_keyring'
2023-09-27 10:00:51 ERROR    Keyring file does not exist. Please try again.
```

**Comment by ilium007** [comment]
With debug logging turned on:

```
root@synology01:~# docker exec -it icloudpd sync-icloud.sh --Initialise

2023-09-27 11:03:57 INFO     ***** boredazfcuk/icloudpd container for icloud_photo_downloader v1.0.635 started *****
2023-09-27 11:03:57 INFO     ***** For support, please go here: https://github.com/boredazfcuk/docker-icloudpd *****
2023-09-27 11:03:57 INFO     Alpine Linux 3.18.3
2023-09-27 11:03:57 INFO     Python version: 3.11.5
2023-09-27 11:03:57 INFO     Loading configuration from: /config/icloudpd.conf
2023-09-27 11:03:57 DEBUG    Running user id: 0
2023-09-27 11:03:57 DEBUG    Running group id: 0
2023-09-27 11:03:57 DEBUG    Local user: user:1000
2023-09-27 11:03:57 DEBUG    Local group: group:1000
2023-09-27 11:03:57 DEBUG    Force GID: false
2023-09-27 11:03:57 DEBUG    LAN IP Address: 192.168.115.1
2023-09-27 11:03:57 DEBUG    Default gateway: 192.168.115.254
2023-09-27 11:03:57 DEBUG    DNS server: 127.0.0.11
2023-09-27 11:03:58 DEBUG    IP address for icloud.com: 17.253.144.10
2023-09-27 11:03:58 DEBUG    Route check to icloud.com successful
2023-09-27 11:03:58 DEBUG    Apple ID: (hidden)
2023-09-27 11:03:58 INFO     Authentication Type: MFA
2023-09-27 11:03:58 DEBUG    Cookie path: /config/(hidden)
2023-09-27 11:03:58 INFO     Cookie expiry notification period: 7
2023-09-27 11:03:58 INFO     Download destination directory: /images
2023-09-27 11:03:58 INFO     Folder structure: {:%Y/%m/%d}
2023-09-27 11:03:58 DEBUG    Directory permissions: 750
2023-09-27 11:03:58 DEBUG    File permissions: 640
2023-09-27 11:03:58 INFO     Synchronisation interval: 86400
2023-09-27 11:03:58 INFO     Synchronisation delay (minutes): 0
2023-09-27 11:03:58 INFO     Set EXIF date/time: true
2023-09-27 11:03:58 INFO     Auto delete: false
2023-09-27 11:03:58 INFO     Delete after download: false
2023-09-27 11:03:58 INFO     Photo size: original
2023-09-27 11:03:58 INFO     Single pass mode: false
2023-09-27 11:03:58 INFO     Skip download check: false
2023-09-27 11:03:58 INFO     Skip live photos: false
2023-09-27 11:03:58 INFO     Number of most recently added photos to download: Download All Photos
2023-09-27 11:03:58 INFO     Downloading photos from album(s): Fortuner
2023-09-27 11:03:58 INFO     Stop downloading when prexisiting files count is: Download All Photos
2023-09-27 11:03:58 INFO     Live photo size: original
2023-09-27 11:03:58 INFO     Skip videos: false
2023-09-27 11:03:58 INFO     Convert HEIC to JPEG: false
2023-09-27 11:03:58 INFO     Downloading from: icloud.com
2023-09-27 11:03:58 INFO     Authentication domain: com
2023-09-27 11:03:58 DEBUG    Nextcloud synchronisation trigger: Disabled
2023-09-27 11:03:58 DEBUG    Nextcloud upload: Disabled
2023-09-27 11:03:58 INFO     Ignore Synology extended attribute directories: Disabled
2023-09-27 11:03:58 INFO     Script launch parameters: --Initialise
2023-09-27 11:03:58 DEBUG    Group, group:1000, already created
2023-09-27 11:03:58 DEBUG    User, user:1000, already created
2023-09-27 11:03:58 DEBUG    Set owner and group on icloudpd temp directory
2023-09-27 11:03:58 DEBUG    Set owner and group on config directory
2023-09-27 11:03:58 DEBUG    Set owner and group on keyring directory
2023-09-27 11:03:58 INFO     Directory is writable: /config/python_keyring/
2023-09-27 11:03:58 INFO     Directory is writable: /home/user/.local/share/
2023-09-27 11:03:58 DEBUG    Configure password
2023-09-27 11:03:58 DEBUG    Adding password to keyring file: /config/python_keyring/keyring_pass.cfg
2023-09-27 11:03:59 DEBUG    Switched to icloudpd: 1.16.0
Enter iCloud password for xxxx@mac.com:
Save password in keyring?  [y/N]: y
Traceback (most recent call last):
  File "/opt/icloudpd_latest/bin/icloud", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/cmdline.py", line 219, in main
    utils.store_password_in_keyring(username, password)
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/utils.py", line 53, in store_password_in_keyring
    return keyring.set_password(
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/keyring/core.py", line 60, in set_password
    get_keyring().set_password(service_name, username, password)
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/keyrings/alt/file_base.py", line 129, in set_password
    self._write_config_value(service, username, password_base64)
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/keyrings/alt/file_base.py", line 137, in _write_config_value
    self._ensure_file_path()
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/keyrings/alt/file_base.py", line 163, in _ensure_file_path
    os.makedirs(storage_root)
  File "<frozen os>", line 225, in makedirs
FileExistsError: [Errno 17] File exists: '/home/user/.local/share/python_keyring'
2023-09-27 11:04:25 ERROR    Keyring file does not exist. Please try again.
```

**Comment by ilium007** [comment]
I don't understand why its looking in `/home/user/.local/share/python_keyring`

**Comment by boredazfcuk** [comment]
This is all correct. The keyring application saves to /home/user/.local/share/python_keyring/keyring_pass.cfg, but as this location is not located on a persistent storage location, the script symlinks /home/user/.local/share/python_keyring to /config/python_keyring.

I've actually re-looked at this to night and figured out that the keyring app checks for the presence of the folder by checking and environment variable. I've set this variable in the Dockerfile and rebuilt the containers, so now it writes directly to /config/python_keyring/ directory.

I don't think this is likely to resolve you issues, but it might do if it's having difficulty navigating the link.

**Comment by ilium007** [comment]
Yesterday I tried taking the complete configure and running on a Debian server and it so worked fine. So this issue for me is only occurring in the Synology NAS running docker. I did run a shell on the container and confirmed the symlink was in place on the Synology so not sure what's different between Debian and the Synology.

**Comment by boredazfcuk** [comment]
Some else describes seeing the same issue with podman containers on MacOS. They switched to Docker and the problem was resolved. My guess is that the same issue exists on Synology NAS devices and was introduced in DSM 7 (possibly 7.2) as people seem to have mentioned it happening after an upgrade.

I stumbled upon this today, and I'm thinking it might be worth a try, because this issue screams "permissions" to me: https://devops.stackexchange.com/questions/11267/permission-denied-within-mounted-volume-inside-podman-container

However, it seems quite dangerous: https://stackoverflow.com/questions/35218194/what-is-z-flag-in-docker-containers-volumes-from-option

I also have a feeling it's because the path mounts are to local paths, rather than dedicated volumes. I created a container for a work project and mounted a volume with ./config:/config and it didn't work properly. When I created a dedicated volume called container_config before creating the container, and mounted that in with container_config:/config it worked fine.
